jib-core/src/main/java/com/google/cloud/tools/jib/builder/ProgressEventDispatcher.java

```diff
@@ -0,0 +1,142 @@
+ * instances with {@link #newChildProducer}.
+ */
+public class ProgressEventDispatcher implements Closeable {
+
+  /**
+   * Creates a new {@link ProgressEventDispatcher} based off an existing {@link
+   * ProgressEventDispatcher}. {@link #create} should only be called once.
+   */
+  @FunctionalInterface
+  public interface Factory {
+
+    /**
+     * Creates the {@link ProgressEventDispatcher} with an associated {@link Allocation}.
+     *
+     * @param description user-facing description of what the allocation represents
+     * @param allocationUnits number of allocation units
+     * @return the new {@link ProgressEventDispatcher}
+     */
+    ProgressEventDispatcher create(String description, long allocationUnits);
+  }
+
+  /**
+   * Creates a new {@link ProgressEventDispatcher} with a root {@link Allocation}.
+   *
+   * @param eventDispatcher the {@link EventDispatcher}
+   * @param description user-facing description of what the allocation represents
+   * @param allocationUnits number of allocation units
+   * @return a new {@link ProgressEventDispatcher}
+   */
+  public static ProgressEventDispatcher newRoot(
+      EventDispatcher eventDispatcher, String description, long allocationUnits) {
+    return newProgressEventDispatcher(
+        eventDispatcher, Allocation.newRoot(description, allocationUnits));
+  }
+
+  /**
+   * Creates a new {@link ProgressEventDispatcher} and dispatches a new {@link ProgressEvent} with
+   * progress 0 for {@code allocation}.
+   *
+   * @param eventDispatcher the {@link EventDispatcher}
+   * @param allocation the {@link Allocation} to manage
+   * @return a new {@link ProgressEventDispatcher}
+   */
+  private static ProgressEventDispatcher newProgressEventDispatcher(
+      EventDispatcher eventDispatcher, Allocation allocation) {
+    return new ProgressEventDispatcher(eventDispatcher, allocation).dispatchProgress(0);
+  }
+
+  private final EventDispatcher eventDispatcher;
+  private final Allocation allocation;
+
+  private long remainingAllocationUnits;
+  private boolean closed = false;
+
+  private ProgressEventDispatcher(EventDispatcher eventDispatcher, Allocation allocation) {
+    this.eventDispatcher = eventDispatcher;
+    this.allocation = allocation;
+
+    remainingAllocationUnits = allocation.getAllocationUnits();
+  }
+
+  /**
+   * Creates a new {@link Factory} for a {@link ProgressEventDispatcher} that manages a child {@link
+   * Allocation}. Since each child {@link Allocation} accounts for 1 allocation unit of its parent,
+   * this method decrements the {@link #remainingAllocationUnits} by {@code 1}.
+   *
+   * @return a new {@link Factory}
+   */
+  public Factory newChildProducer() {
+    decrementRemainingAllocationUnits(1);
+
+    return new Factory() {
+
+      private boolean used = false;
+
+      @Override
+      public ProgressEventDispatcher create(String description, long allocationUnits) {
+        Preconditions.checkState(!used);
```

Comment:
Just keep in mind that generally speaking, it's better if restrictions can be enforced at the API level (the API prevents mis-usage from the get-go), so try to go that way as much as possible if achievable.

Anyways, if `create()` should be called once, is it not possible to somehow remove calling `newChildProducer().create()` in a chain? I see you call `someDispatcher.newChildProducer()` for passing factory to another method and the callee calls `create()` there, and just from the technical viewpoint, you can pass `someDispatcher` to the callee (which I think is basically the previous code before this PR). I know the `close()` method matters with `try` now, but isn't it that any callees should complete before going out of the `try` block anyway? Trying to understand why we need to have the factory and then calling `create()`.

Reply:
The factory is there to first decrement the `remainingAllocationUnits` and passing that `Factory` to the child `Step`. The child `Step` does not create its own `ProgressEventDispatcher` with the `Factory` until it actually runs since creation of the `ProgressEventDispatcher` means to also send a `ProgressEvent` with 0 progress to add that step as an unfinished step (we don't want to add it before it starts running).

An alternative could be to not have the `Factory` but rather just pass the child `ProgressEventDispatcher`s and have the `try-with-resource` try on a `ProgressEventDispatcher.Manager` that on creation sends a `ProgressEvent` with 0 progress and on close does what `ProgressEventDispatcher#close` does now.

Reply:
The `dispatchProgress(0)` is called in `newProgressEventDispatcher` on line 77.